File: doc/src/codes/Users_Manual/advanced_examples/custom_modulation_0.py
import numpy as np
from pyautd3 import EmitIntensity, SamplingConfiguration
from pyautd3.modulation import Modulation


class Burst(Modulation):
    _length: int

    def __init__(self, length: int = 4000):
        super().__init__(SamplingConfiguration.from_frequency(4e3))
        self._length = length

    def calc(self):
        buf = np.array([EmitIntensity.minimum()] * self._length)
        buf[0] = EmitIntensity.maximum()
        return buf

File: doc/src/codes/Users_Manual/advanced_examples/custom_modulation_0.cpp
#include "autd3.hpp"

class BurstModulation final : public autd3::Modulation {
 public:
  std::vector<autd3::EmitIntensity> calc() const override {
    std::vector buffer(_buf_size, autd3::EmitIntensity::minimum());
    buffer[_buf_size - 1] = autd3::EmitIntensity::maximum();
    return buffer;
  }

  explicit BurstModulation(const size_t buf_size = 4000) noexcept
      : autd3::Modulation(autd3::SamplingConfiguration::from_frequency(4e3)), _buf_size(buf_size) {}

 private:
  size_t _buf_size;
};